### PR TITLE
get organization and user data from in lms

### PR DESCRIPTION
**Pull request by zubair-arbi**
ECOM-3282
@awais786 @ahsan-ul-haq @tasawernawaz 

**WIP**

* Create settings variables required for communication with the `organizations` API in `LMS`
* Add `get_organization` method to get an organization from edx rest api client with the provided `organization_key`

**Remaining changes**
* Create settings variables required for communication with the `users` API in `LMS`
* Add `get_user` method to get a user data from edx rest api client with the provided `username`
* Add tests for data retrieval from `organizations` api
* Add tests for data retrieval from`users` api

### CODE REVIEW

**Comment by zubair-arbi** [comment]
fyi: @jimabramson @clintonb

File: credentials/apps/credentials/utils.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 from urlparse import urljoin
 
 import jwt
+import requests
```
Comment (clintonb) [inline credentials/apps/credentials/utils.py]: This is not used.

File: credentials/settings/base.py [hunk 2]
```diff
@@ -213,8 +213,18 @@
 PROGRAMS_API_URL = '/api/v1/'
 # Specified in seconds. Enable caching by setting this to a value greater than 0.
 PROGRAMS_CACHE_TTL = 0
-# Credentials service user in Programs service
-PROGRAMS_SERVICE_USER = 'credentials_service_user'
+
+# Absolute URL used to construct URLs pointing to the LMS
+LMS_URL_ROOT = 'http://127.0.0.1:8000'
+# Organizations API url in edx-platform
+ORGANIZATIONS_API_URL = '/api/organizations/v0/'
```
Comment (clintonb) [inline credentials/settings/base.py]: This should be a complete URL, not a path. That way, if we ever move the service to an IDA, your code does not need to change. I would do the same for `PROGRAMS_API_URL` as well.

File: credentials/apps/credentials/utils.py [hunk 3]
```diff
@@ -122,3 +183,24 @@ def _get_service_user_api_client(api_url, service_username):
...
+    Arguments:
+        api_url (str): Absolute url of the api
+
+    """
+    try:
+        api_client = EdxRestApiClient(
+            get_organizations_api_url(),
+            oauth_access_token=settings.CREDENTIALS_OAUTH_ACCESS_TOKEN
+        )
+    except Exception:  # pylint: disable=broad-except
```
Comment (clintonb) [inline credentials/apps/credentials/utils.py]: When/why would an exception be raised? Are there less-broad exceptions you can catch?

File: credentials/apps/credentials/utils.py [hunk 4]
```diff
@@ -122,3 +183,24 @@ def _get_service_user_api_client(api_url, service_username):
...
+    Helper method to get edx rest api client for the credentials service user
+    which is present on the LMS.
+
+    Arguments:
+        api_url (str): Absolute url of the api
+
+    """
+    try:
+        api_client = EdxRestApiClient(
+            get_organizations_api_url(),
```
Comment (clintonb) [inline credentials/apps/credentials/utils.py]: This should be `api_url`. Remove the `branches` directive in .travis.yml so that tests are run on this PR: https://github.com/edx/credentials/blob/master/.travis.yml#L22-L24.

File: credentials/apps/credentials/utils.py [hunk 5]
```diff
@@ -122,3 +183,24 @@ def _get_service_user_api_client(api_url, service_username):
         return
 
     return api_client
+
+
+def _get_lms_service_user_api_client(api_url):
```
Comment (clintonb) [inline credentials/apps/credentials/utils.py]: Why is this necessary? Do you plan to use it elsewhere?

File: credentials/settings/base.py [hunk 6]
```diff
@@ -213,8 +213,18 @@
...
+# Specified in seconds. Enable caching by setting this to a value greater than 0.
+ORGANIZATIONS_CACHE_TTL = 0
```
Comment (clintonb) [inline credentials/settings/base.py]: This should default to a sane production default (e.g. 30 minutes). You can set it to a lower development default in local.py.

File: credentials/apps/credentials/utils.py [hunk 7]
```diff
@@ -77,6 +79,49 @@ def get_programs():
...
+
+    Arguments:
+        organization_key (str): Unique key of the organization for retrieval
+
+    Returns:
+        dict, representing organization data returned by the LMS.
+    """
+    no_organization = {}
+
+    # Bypass caching if the programs cache ttl is not set.
```
Comment (clintonb) [inline credentials/apps/credentials/utils.py]: Remove this to eliminate complexity. Django already supports what you're doing: https://docs.djangoproject.com/en/1.9/topics/cache/#basic-usage.

File: credentials/apps/credentials/utils.py [hunk 8]
```diff
@@ -77,6 +79,49 @@ def get_programs():
...
+    use_cache = False
+    if settings.ORGANIZATIONS_CACHE_TTL > 0:
+        use_cache = True
+
+    if use_cache:
+        cached = cache.get(ORGANIZATIONS_CACHE_KEY)
```
Comment (clintonb) [inline credentials/apps/credentials/utils.py]: Why are all organizations being cached as a single key? That seems unnecessarily complicated. Consider caching one value per organization.

```python
# TODO Ensure the cache keys are ASCII and less than 250 characters
cache_key = 'organization_{}'.format(hash(organization_key))

organization = cache.get(cache_key)

if not organization:
  # TODO Attempt to retrieve info from API

  # Do not cache empty values
  if organization:
    cache.set(cache_key, organization)

return organization
```

File: credentials/apps/credentials/utils.py (hunk 3 above)
Comment (zubair-arbi) [inline credentials/apps/credentials/utils.py]: I have seen `ConnectionError`, `HttpClientError` and `AttributeError`. Handling this broad exception is consistent with the programs code.

**Comment by zubair-arbi** [comment]
Closing this PR in favor of https://github.com/edx/credentials/pull/34
